includes/class-wc-payments-upe-checkout.php

```diff
@@ -0,0 +1,299 @@
+			if ( ! empty( $prepared_customer_data ) ) {
+				wp_localize_script( 'wcpay-upe-checkout', 'wcpayCustomerData', $prepared_customer_data );
+			}
+
+			wp_enqueue_style(
+				'wcpay-upe-checkout',
+				plugins_url( 'dist/checkout.css', WCPAY_PLUGIN_FILE ),
+				[],
+				WC_Payments::get_file_version( 'dist/checkout.css' )
+			);
+
+			// Output the form HTML.
+			?>
+			<?php if ( ! empty( $this->gateway->get_description() ) ) : ?>
+				<p><?php echo wp_kses_post( $this->gateway->get_description() ); ?></p>
+			<?php endif; ?>
+
+			<?php if ( $this->gateway->is_in_test_mode() ) : ?>
+				<p class="testmode-info">
+					<?php
+						$testing_instructions = $this->gateway->payment_method->get_testing_instructions();
+					if ( false !== $testing_instructions ) {
+						echo WC_Payments_Utils::esc_interpolated_html(
+							/* translators: link to Stripe testing page */
+							$testing_instructions,
+							[
+								'strong' => '<strong>',
+								'a'      => '<a href="https://woocommerce.com/document/payments/testing/#test-cards" target="_blank">',
+							]
+						);
+					}
+					?>
+				</p>
+			<?php endif; ?>
+
+			<?php
+
+			if ( $display_tokenization ) {
+				$this->gateway->tokenization_script();
+				$this->gateway->saved_payment_methods();
+			}
+			?>
+
+			<fieldset id="wc-<?php echo esc_attr( $this->gateway->id ); ?>-upe-form" class="wc-upe-form wc-payment-form">
+				<div class="wcpay-upe-element" data-payment-method-type="<?php echo esc_attr( $this->gateway->get_selected_stripe_payment_type_id() ); ?>"></div>
+				<?php
+					$is_enabled_for_saved_payments = $this->gateway->is_enabled_for_saved_payments();
+				if ( $this->gateway->is_saved_cards_enabled() && $is_enabled_for_saved_payments ) {
+					$force_save_payment = ( $display_tokenization && ! apply_filters( 'wc_payments_display_save_payment_method_checkbox', $display_tokenization ) ) || is_add_payment_method_page();
+					if ( is_user_logged_in() || $force_save_payment ) {
+						$this->gateway->save_payment_method_checkbox( $force_save_payment );
+					}
+				}
+				?>
+
+			</fieldset>
+
+			<?php if ( Fraud_Prevention_Service::get_instance()->is_enabled() ) : ?>
+				<input type="hidden" name="wcpay-fraud-prevention-token" value="<?php echo esc_attr( Fraud_Prevention_Service::get_instance()->get_token() ); ?>">
+			<?php endif; ?>
+
+			<?php
+
+			do_action( 'wcpay_payment_fields_upe', $this->gateway->id );
+
+		} catch ( \Exception $e ) {
+			// Output the error message.
+			Logger::log( 'Error: ' . $e->getMessage() );
+			?>
+			<div>
+				<?php
+				echo esc_html__( 'An error was encountered when preparing the payment form. Please try again later.', 'woocommerce-payments' );
+				?>
+			</div>
+			<?php
+		}
+	}
+
+	/**
+	 * Set gateway
+	 *
+	 * @param string $gateway_id UPE gateway ID.
+	 * @return void
+	 */
+	public function set_gateway( $gateway_id ) {
+		if ( null !== $gateway_id ) {
+			$this->gateway = $this->gateway_map[ $gateway_id ];
```

Comment:
```suggestion
			$this->gateway = WC_Payments::get_payment_gateway_by_id( $gateway_id );
```
Could be safer to call this function here, yeah?

Reply:
```suggestion
			$this->gateway = $this->gateway->wc_payments_get_payment_method_by_id( $gateway_id );
```
There's this one as well, if you'd prefer using the private gateway, instead of the static function; though this function is merely a wrapper around the same static function anyways.

Reply:
Great idea! I also removed the `$gateway_map` from the checkout class. Thanks!